drivers: i2s: max32: Driver with TX/RX Support

cx-anuj-pathak opened the pull request:
### Description
Adds I2S driver for MAX32655 MCU. It was only tested on MAX32655FTHR board using `samples/i2s/echo` sample.
Theoretically it should work on most MAS32 MCU (except MAX32650/MAX32660 which uses SPIMSS peripheral), given same MXC_I2S_* APIs and config structure are used internally. But due to lack of hardware it is not tested and dts of other mcu/board is also not updated.

### Testing 
- checkout to branch and `west update`
- build using `west build -b max32655fthr/max32655/m4 ..\zephyr\samples\drivers\i2s\echo\`
- flash using `openocd.exe -f interface/cmsis-dap.cfg -f target/max32655.cfg -c "init" -c "reset" -c "halt" -c "flash write_image erase build/zephyr/zephyr.hex" -c "reset" -c "exit"` (note: use openocd from msdk)
- speaking into max32655fthr will be echoed to the speaker through the jack.

cx-anuj-pathak committed:
dts: i2s: add i2s node for max32655fthr

- add basic dma only i2s dts binding for max32-i2s
- add i2s node with default config to max32655.dtsi
- add i2s pin definition for max32655fthr board as per spec

Signed-off-by: Anuj Pathak <anuj@croxel.com>

cx-anuj-pathak committed:
drivers: i2s: add i2s driver for max32 mcu

add necessary build files and i2s driver with tx/rx support

Signed-off-by: Anuj Pathak <anuj@croxel.com>

cx-anuj-pathak committed:
boards: max32655fthr: enable i2c1

MAX32655 uses I2C1 to control on board PMIC and Audio Codec
this commit enable it as per board spec

Signed-off-by: Anuj Pathak <anuj@croxel.com>

cx-anuj-pathak committed:
samples: drivers: i2s: echo: Add MAX9867 codec support

Add MAX9867 codec initialization support to the sample
to later use for MAX32655FTHR support

Signed-off-by: Anuj Pathak <anuj@croxel.com>

bricle commented:
@cx-anuj-pathak Thanks for your excellent work, any plans to move this forward? thanks.

cx-anuj-pathak commented:
> @cx-anuj-pathak Thanks for your excellent work, any plans to move this forward? thanks.

@bricle Yes, I'm active and will iterate over it as soon as dependencies in adi_hal projects are merged. I guess it got paused due to 4.2 release, and hopeful it will resume soon.

cx-anuj-pathak commented:
> > > > @cx-anuj-pathak Thanks for your excellent work, any plans to move this forward? thanks.
> > > 
> > > 
> > > @bricle Yes, I'm active and will iterate over it as soon as dependencies in adi_hal projects are merged. I guess it got paused due to 4.2 release, and hopeful it will resume soon.
> > > Right it forces custom branch in manifest, and causes PR Metadata check fail. That can only be resolved once adi_hal PRs are merged.
> > 
> > 
> > Very glad to hear this! But it seems no one have noticed these two PRs, maybe you could try to directly ping some regular reviewers from ADI to come into the spotlight. Thanks again for your reply.
> 
> Thanks for the PR. The adi_hal module is a downstream fork of https://github.com/analogdevicesinc/msdk. Could you please send the hal changes upstream?

Filed PRs pointing to msdk repo at
https://github.com/analogdevicesinc/msdk/pull/1425
https://github.com/analogdevicesinc/msdk/pull/1426

let me know further coarse of actions

cx-anuj-pathak commented:
@MaureenHelm PR to msdk are merged, let me know further coarse of actions

cx-anuj-pathak committed:
samples: drivers: i2s: echo: add MAX32655 support

- reduced block length from 100ms to 33.33ms to compile for low ram
- add compile time conditional option to slave mode
- add overlay file for max32655fthr

Signed-off-by: Anuj Pathak <anuj@croxel.com>

ttmut commented on samples/drivers/i2s/echo/src/codec.c:
```diff
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2021 Nordic Semiconductor ASA
+ * Copyright (c) 2021 Croxel Inc
```
```suggestion
 * Copyright (c) 2025 Croxel Inc
```
2025?

ttmut reviewed (changes requested):
Thanks for putting this together. I’ve added a few comments that you might find helpful.

cx-anuj-pathak commented on dts/bindings/i2s/adi,max32-i2s.yaml:
```diff
@@ -0,0 +1,33 @@
+    default: 12288000
+    description: |
+      Frequency of the I2S clock in Hz.
+      This clock is used to derive the bit clock and frame clock.
```
Added a comment, let me know if it is sufficient.
Or 
I can remove the default value and then put a build_assert in code to make sure this value is set for MAX78000/2 and MAX32690

cx-anuj-pathak commented on samples/drivers/i2s/echo/src/main.c:
```diff
@@ -21,12 +21,18 @@
 #define I2S_TX_NODE  DT_NODELABEL(i2s_tx)
 #endif
 
+#if CONFIG_BOARD_MAX32655FTHR_MAX32655_M4
```
for some reason CONFIG_SRAM_SIZE  is getting 48 value in the .config file generated. may be a bug somewhere?
I tried to look for some other portable ways using autogenerated DTS macros but could found any.

cx-anuj-pathak commented on samples/drivers/i2s/echo/dts/bindings/adi,max9867.yaml:
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) 2025 Croxel Inc.
```
Added that copyright too, since it is based on that

ttmut commented on include/zephyr/dt-bindings/dma/max32655_dma.h:
```diff
@@ -28,6 +28,7 @@
 #define MAX32_DMA_SLOT_UART3_TX 0x3CU
-#define MAX32_DMA_SLOT_I2S      0x3EU
+#define MAX32_DMA_SLOT_I2S_TX   0x3EU
+#define MAX32_DMA_SLOT_I2S_RX   0x1EU
```
Move `MAX32_DMA_SLOT_I2S_RX` up just below `MAX32_DMA_SLOT_UART3_RX` to keep the incremental order.

MaureenHelm commented on drivers/i2s/i2s_max32.c:
```diff
@@ -0,0 +1,774 @@
+	return 0;
+}
+
+static const struct i2s_driver_api i2s_max32_driver_api = {
```
Use `DEVICE_API`

MaureenHelm commented on dts/bindings/i2s/adi,max32-i2s.yaml:
```diff
@@ -0,0 +1,34 @@
+    description: |
+      Names of DMA channels for RX and TX. Must be provided in the following order:
+      - rx
+      - tx
```
Ditto

MaureenHelm commented on boards/adi/max32655fthr/max32655fthr_max32655_m4.dts:
```diff
@@ -144,6 +144,12 @@
 
+&i2c1 {
+	status = "okay";
+	pinctrl-0 = <&i2c1_scl_p0_16 &i2c1_sda_p0_17>;
```
Do these pins need to operate at 1.8 or 3.3V? If 3.3V, then you need to set `MAX32_VSEL_VDDIOH`

MaureenHelm reviewed (changes requested):
Some minor things and needs a rebase, but overall looks great. Thanks for the PR!

cx-anuj-pathak commented on dts/bindings/i2s/adi,max32-i2s.yaml:
```diff
@@ -0,0 +1,34 @@
+    description: |
+      DMA channels for RX and TX. Two channels must be provided in the following order:
+      - RX DMA channel
+      - TX DMA channel
```
Not really, removed

cx-anuj-pathak commented on samples/drivers/i2s/echo/boards/max32655fthr_max32655_m4.overlay:
```diff
@@ -0,0 +1,16 @@
+		compatible = "adi,max9867";
+		reg = <0x18>;
+	};
+};
```
```
i2s_rxtx: &i2s0 {
	status = "okay";
};
```
this is out of `max32655fthr_max32655_m4.dts` to allow application to enable I2S only if used, as per following kconfig
```
config I2S_MAX32
	bool "Analog Devices MAX32 series I2S driver"
	default y
	depends on DT_HAS_ADI_MAX32_I2S_ENABLED
	select DMA
	help
	  Enable support for Analog Devices MAX32 series I2S driver.
```

```
&i2c1 {
	max9867: max9867@18 {
		compatible = "adi,max9867";
		reg = <0x18>;
	};
};
```
This is also not in `max32655fthr_max32655_m4.dts` because we are using out-of-tree minimalistic code to initialize max9867

MaureenHelm replied:
We can still define the node at the board level. The devicetree describes the hardware, not the software.

cx-anuj-pathak commented:
@anangl friendly reminder.
